services/src/main/java/org/keycloak/keys/loader/PublicKeyStorageManager.java

```diff
@@ -46,6 +47,19 @@ public static PublicKey getClientPublicKey(KeycloakSession session, ClientModel
         return keyStorage.getPublicKey(modelKey, kid, loader);
     }
 
+    // KEYCLOAK-8460 client signed signature verification
+    // key is wrapped by KeyWrapper in order to use refactored token signature verification mechanism
+    public static KeyWrapper getClientPublicKeyWrapper(KeycloakSession session, ClientModel client, JWSInput input) {
+        String kid = input.getHeader().getKeyId();
```

Comment:
jwks_uri is the ideal location to fetch keys and it should advertise the kids for the keys, so it's just a matter of picking the correct key. That allows the clients to rotate keys as well without breaking.

If public keys for client is uploaded directly using a PEM or something that should really allow specifying the kid as well.

I don't have that much knowledge about how this particular feature works though. Perhaps @mposolda can comment here?

Reply:
So I will try to refactor client public key loading to return KeyWrapper rather than PublicKey.

I will consider its specification and design at first, and comment here afterwards. I would be happy if you comment on it.

Reply:
So I will try to refactor client public key loading to return KeyWrapper rather than PublicKey.

I will consider its specification and design at first, and comment here afterwards. I would be happy if you comment on it.

Reply:
So I will try to refactor client public key loading to return KeyWrapper rather than PublicKey.

I will consider its specification and design at first, and comment here afterwards. I would be happy if you comment on it.